### DISCUSSION TITLE
Intergration of historical(typical) traffic data from Mapbox

### DISCUSSION THREAD

**Discussion by xlqian** [valhalla/valhalla#5391] [Q&A]
Hi, 

we've recently started working on the integration of historical traffic data from Mapbox and here is [the format of the data](https://docs.mapbox.com/data/traffic/guides/data/#typical-file-contents).  I've been browsing the repo around and couldn't find an elegant solution.

Although We managed to run this project https://github.com/alinmindroc/valhalla_traffic_poc for a better understading of how traffic works in Valhalla, but somehow we hit a snag  when it comes to mapping the data from Mapbox to edges in Valhalla, since,  instead of a osm_way_id, the data provided by Mapbox comes with only "osm_start_node"  and "osm_start_node" to identify the portion of way. 

Indeed, there're numerous ways to identify the osm_way_id by using  "osm_start_node"  and "osm_start_node". Here I'm wondering if we can make good use of [ these intermediate bin files ](https://github.com/valhalla/valhalla/blob/c411feb309618c956b5fa7170d3845fb9fa75a06/src/mjolnir/util.cc#L275) to construct a map `(osm_start_node, osm_start_node) vs valhalla_edges` (just like `valhalla_ways_to_edges` does)    

Or these is a better way to achieve this?  

thank you

**Comment by kevinkreiser** [comment]
i think openlr is the best way. its quite straight forward with valhalla to write a little tool that maps openlr segments to a sequence of valhalla edges. this also has general purpose use for other data-set-agnostic overlays that other companies might produce such as live traffic data sets. if we could add tool or at least c++ library api's that take a given openlr and use the basic features of valhalla to "map match" them to the graph that should be sufficient. then we dont need to keep large auxilliary data sets around (the bins) and we are also robust to matching datasets that may not be osm.

side note, i am biased on this as i have written this type of code before.

**Comment by xlqian** [comment]
hi @kevinkreiser , thank you so much for the speedy answer. Let me have a close look into openlr. 

cheers

**Comment by kevinkreiser** [comment]
@xlqian no worries and also please let me know if you want to bat around ideas. id be happy to sketch up a rough outline of how to do it if you are interested!

**Comment by xlqian** [comment]
Totally, I admit that openlr is entirely new for me. I would love to have some guide and do a PR for that matter!

**Comment by kevinkreiser** [comment]
ok cool i can write up the gist of how to do that today sometime, might be tomorrow for you by the time i get it (assume you are still in CET)

**Comment by kevinkreiser** [comment]
ran out of time tonight, will do tomorrow morning

**Comment by kevinkreiser** [comment]
ok ideally i think we need to build some tooling. here are my thoughts from previous experience:

## input

an openlr record, binary. we have already openlr.h which can read these into a usable in memory struct that we can operate on

## output

for a given openlr record you get back a sequence of contiguous edges in the graph, and with each edge a begin and end percent, very much like the internal output we get from the map matching algorithm (called EdgeSegments in that part of the code)

## openlr router

you can kind of think of openlr as a route request specification/format. its basiaclly just a sequence of points with metadata on each point that dictates to the router both the "shape" and the "types of edges" that should be used in the route finding. this means we can decompose an openlr record pretty much into a series of routing requests. i have previously written code to do this in c++ but since it is just routing requests you could presumably even do it in python using our higher level apis if you wanted to. i think i would still prefer it in c++ so that it doesnt preclude us doing it in other novel ways but let me know what you think.

TODO: ill update this issue with an implementation outline and details of how i would propose we do path finding with the openlr segment. at a high level we'll simply be converting it into a list of locations, using loki::search and using thor::astar. bare with me

### python-c++-ish pseudo code

```python
# ranking penalization function, returns a value of 0 1 or 2
rank:
  frc mismatch + (motorway mismatch || oneway mismatch || slip road mismatch)

# make a special (trivial) costing to use inline with the algorithm, doesnt need to be added to real costings but should follow their interface so we can pass to astar
# uses vehicular access only
# allowed methods for predecessors disallow uturns and shortcut edges
# allowed method for nodes allows all
# allowed method for edge disallows shortcuts and edges with rank > 1
# edge cost is just distance

#piecewise openlr processing by lrp
for each pair of lrps:
  find a route for this pair:
  
    # basic loki search, very specific numbers below come from openlr specs accuracy (its a binary packed struct like we use throughout valhalla with fixed precision, look at the openlr parser and you'll see)
    loki search both lrps:
      search cut off 20m
      node snap 2m
      heading tolerance 34 degrees
      flip last lrp heading by 180
    
    # rank search results using ranking function above
    for each loki result:
      for each edge candidate:
        if rank > 1:
          remove candidate
        if candidate is start lrp but is at end of edge and beginning of next edge is there:
          remove candidate
        if candidate is end lrp but is at beginning of edge and end of next edge is there:
          remove candidate
      sort remaining candidates by rank (ascending, 0 is the best score)
    
    # find trivial paths (single edge routes, ie same edge candidate on both lrps)
    trivial paths:
      for all pairs of candidates:
        if they are the same graphid (trivial route) and they meet the length requirement keep them
      sort them by rank of first candidate
    if the top ranked candidate of both lrps makes a trivial route within length tolerance:
      return that as the best path
      
    # routing for less trivial openlr segments that span multiple edges
    find and check the route using just the best ranked candidate of each lrp (ie src_lrp.candidates[0] and dst_lrp.candidates[0]
    if that met the length requirement:
      return that route
    try again with all candidates
    if that met the length requirement:
      return that route
    if there were trivial paths:
      return top ranked trivial path
    
  # matching lrp offsets
  if first lrps and poff:
    then trim beginning
  if last lrps and noff
    trim end
    
# put the pieces together again for the whole openlr
merge matches through intermediate lrps (concat removing dups)
```

**Comment by xlqian** [comment]
thank you SO much for taking your time to respond to us. 

> you can kind of think of openlr as a route request specification/format. ... this means we can decompose an openlr record pretty much into a series of routing requests.

~i don't really get this part, could you please elaborate a little bit more?  How does "routing requests" have something to do with the integration of historical speed into Valhalla?~


The version that I've imagined was use "map match" to find edges corresponding to the given openlr binary and then generate a "valhalla compliant" traffic data that could be injected into valhalla_tiles by using `valhalla_add_predicted_traffic`. 

My solution sounds definitely more naive and tedious...

**Comment by kevinkreiser** [comment]
ive added highlights that i can remember from the couple of times ive implemented this. its all pretty much based on what the openlr spec says to do but condensed into valhalla parlance a little bit. let me know if it makes sense. i think once we have this we can pretty quickly compute a mapping of openlr segments to valhalla edges and percentages along, and use that to assign records to each edge to feed into `valhalla_add_predicted_traffic`. now there will be issues where two openlr segments match a given edge. in this case we'll need to do a tie breaker where the one with the most coverage wins. another option would be to blend the two records together to make a new record but thats probably not worth it for the first go round

**Comment by xlqian** [comment]
OK, after reading some documents about openLR, I finally get your idea!!  I just downloaded the sample data from Mapbox which is about 2.4GB... I'm wondering, by doing a path finding for each segment of each openlr record,  wouldn't it take ages to treat all rows?    

I have another question, since openLR will give a sequence of coordinates, is it viable to find the edges by calling map match in Meili?

**Comment by kevinkreiser** [comment]
> I have another question, since openLR will give a sequence of coordinates, is it viable to find the edges by calling map match in Meili?

at the current moment, it is not viable as meili doesnt have the controls to allow for proper matching (ie matching > 99% of openlrs). if you read the pseudo code above you'll see a few things that just arent there in the map matching api:

1. heading filter when doing candidate search
2. frc fow matching when doing candidate search and route
3. node snapping (openlr is supposed to just route between nodes)
4. specialized/simplified costing

that said, if you just used map matching in meili, you might get a pretty decent usability rate. i wouldnt be surprised if even 90% of the openlrs had clean matches. this might be good enough for you and if that is the case then you could certainly just convert them to map match requests against valhalla_service with trace_attributes. for a longer term more robust solution though id highly advocate for the above strategy as i have seen it work in practice and the openlr docs recommend such an approach

**Comment by kevinkreiser** [comment]
> wouldn't it take ages to treat all rows?

i assume the data you got had both the openlr as well as the historical traffic data that goes along with it? the bulk of that data is the historical data not the openlr part that you need to route with. also the openlrs are almost all trival routes (or should be) and even when they are not they do not require large expansions, they are typically 1 edge routes with the occasional 2 3 4 edge route in there. if you parallelize the processing it should be fast to compute all the mappings of openlr to graph ids. and by fast i mean like less than an hour probably significantly less if you have a decent number of cores

**Comment by xlqian** [comment]
I'm working on this, I'll very likely ask you more questions about it!
